Source/UIKit additions/SVGKFastImageView.m

```diff
@@ -24,31 +24,35 @@ @implementation SVGKFastImageView
 @synthesize timeIntervalForLastReRenderOfSVGFromMemory = _timeIntervalForLastReRenderOfSVGFromMemory;
 
 #if TEMPORARY_WARNING_FOR_APPLES_BROKEN_RENDERINCONTEXT_METHOD
-+(BOOL) svgImageHasNoGradients:(SVGKImage*) image
++(BOOL)svgImageHasNoGradients:(SVGKImage*)image
 {
-	return [self svgElementAndDescendentsHaveNoGradients:image.DOMTree];
+    return [self svgElementAndDescendents:image.DOMTree haveNoClass:[SVGGradientElement class]];
 }
 
-+(BOOL) svgElementAndDescendentsHaveNoGradients:(SVGElement*) element
++ (BOOL)svgImageHasNoText:(SVGKImage*)image
 {
-	if( [element isKindOfClass:[SVGGradientElement class]])
-		return FALSE;
-	else
-	{
-		for( Node* n in element.childNodes )
-		{
-			if( [n isKindOfClass:[SVGElement class]])
-			{
-				if( [self svgElementAndDescendentsHaveNoGradients:(SVGElement*)n])
-					;
-				else
-					return FALSE;
-			}
-				
-		}
-	}
-	
-	return TRUE;
+    return [self svgElementAndDescendents:image.DOMTree haveNoClass:[SVGTextElement class]];
+}
+
++ (BOOL)svgElementAndDescendents:(SVGElement*)element haveNoClass:(Class) theClass
```

Comment:
This seems a mess. This appears to be changing it to require pass in a parameter of a "class that the caller needs to know is a magic class that they MUST pass in the correct class (for no reason) because the code needs it or else the code breaks". WTF?

We have a FIVE LINE COMMENT in the original source explaining why this would be the wrong thing to do and is completely unnecessary --- unless the original comment is a lie, this change appears to ignore the comment completely, delete it, and then substitute meaningless illegible code for no purpose!

(I do not know if the original comment is true, or if it was true and Apple's implenmentation has now changed, or if it was never true. But for all three variations, this code change seems a bad one. If the original comment is no longer true, we can simplify the method name and delete the comment. But I don't see why anyone would want to extend the method, make it harder to use, and without explanation)

Reply:
I just copy the comment from [MaddTheSane's fork - SVGKFastImageView](https://github.com/MaddTheSane/SVGKit/blob/master/Source/UIKit%20additions/SVGKFastImageView.h#L66-L68). Maybe he know the reason Apple's layer implementation on macOS need this ?

I can try to delete this and see the result.

Reply:
@MaddTheSane What's the problem that macOS's `SVGFastImageView` rendering `<text>` element ? If it's only something about quartz coordinate system (flip on macOS, non-flip on iOS), it's easy to fix by using [NSView.isFlipped](https://developer.apple.com/documentation/appkit/nsview/1483532-flipped)

Reply:
IIRC, if rendered using CoreAnimation, text would be right side up. But if rendered without CoreAnimation, it would be upside-down.

Reply:
@MaddTheSane You means, if I use `SVGFastImageView` as layer-backed view (wantsLayer = YES), the render result is flipped ? But if I don't use layer (wantsLayer = NO), the render is correct ? Sounds strange.

I'll try with the branch and using a demo. If so, I guess it's not hard to fix. (For example, we can change something in the `SVGKLayer` parsing and add a transform specially for macOS)

Reply:
I found it's a bug because of the implementation of `SVGTextElement`. I'll try to fix it on macOS

Reply:
@MaddTheSane My fault. It seems the Apple's bug in `-[CALayer renderInContext:]` for `CATextLayer`. Best way to solve it, it to turn on `USE_SUBLAYERS_INSTEAD_OF_BLIT = 1`. Which will use `sublayers` and it works like `SVGKLayeredImageView`.

---

If you still want to use it as normal drawRect view. Go on and see below:

It seems that `CATextLayer` render in the context will flip the content (Text String) on macOS. One hack way, it to override the `contentsAreFlipped` property and return YES. However, it needs a subclass.

Another hack way, we can use the `contentsTransform` (Used by JPEG images, which flip when EXIF orientation is upside down). Here is the simple way of code.

```objectivec
// SVGKFastImageView.m

            CALayer *layerTree = self.image.CALayerTree;
#if TEMPORARY_WARNING_FOR_APPLES_BROKEN_RENDERINCONTEXT_METHOD
            [self preprocessLayerTree:layerTree];
#endif
	    [layerTree renderInContext:context];
#if TEMPORARY_WARNING_FOR_APPLES_BROKEN_RENDERINCONTEXT_METHOD
            [self postProcessLayerTree:layerTree];
#endif


- (void)preprocessLayerTree:(CALayer *)layerTree {
#if SVGKIT_MAC
    if ([layerTree isKindOfClass:[CATextLayer class]]) {
        // Hack to apply flip for text content
        NSAffineTransform *flip = [NSAffineTransform transform];
        [flip scaleXBy:1.0 yBy:-1.0];
        [layerTree setValue:flip forKey:@"contentsTransform"];
    }
#endif
    for (CALayer *layer in layerTree.sublayers) {
        [self preprocessLayerTree:layer];
    }
}

- (void)postProcessLayerTree:(CALayer *)layerTree {
#if SVGKIT_MAC
    if ([layerTree isKindOfClass:[CATextLayer class]]) {
        // Hack to recover flip for text content
        NSAffineTransform *flip = [NSAffineTransform transform];
        [flip scaleXBy:1.0 yBy:1.0];
        [layerTree setValue:flip forKey:@"contentsTransform"];
    }
#endif
    for (CALayer *layer in layerTree.sublayers) {
        [self postProcessLayerTree:layer];
    }
}
```